finetuner/tuner/callback/training_checkpoint.py

```diff
@@ -0,0 +1,143 @@
+
+from .base import BaseCallback
+
+if TYPE_CHECKING:
+    from ..base import BaseTuner
+
+
+class TrainingCheckpoint(BaseCallback):
+    """
+    Callback to save model at every epoch or the last K epochs
+    """
+
+    def __init__(self, save_dir: str, last_k_epochs: int = 1):
+        """
+        :param save_dir: string, path to save the model file.
+        :param last_k_epochs: this parameter is an integer. Only the most
+            recent k checkpoints will be kept. Older checkpoints are deleted
+        """
+        self._logger = JinaLogger(self.__class__.__name__)
+        self._save_dir = save_dir
+        self._last_k_epochs = last_k_epochs
+        self._saved_checkpoints = []
+        if not save_dir:
+            raise ValueError(
+                '``save_dir`` parameter is mandatory. Pass it in parameters'
+            )
+
+    def on_epoch_end(self, tuner: 'BaseTuner'):
+        """
+        Called at the end of the training epoch.
+        """
+        self._save_model_framework(tuner)
+        self._logger.logger.info(
+            f'Model trained for {tuner.state.epoch+1} epochs is saved!'
+        )
+        if self._last_k_epochs:
+            if len(self._saved_checkpoints) > self._last_k_epochs:
+                if os.path.isfile(self._saved_checkpoints[0]):
+                    os.remove(self._saved_checkpoints[0])
+                else:
+                    shutil.rmtree(self._saved_checkpoints[0])
+                self._saved_checkpoints.pop(0)
+                self._logger.logger.info(
+                    f'Model trained for {tuner.state.epoch+1-self._last_k_epochs} epochs is deleted!'
+                )
+
+    def _save_model_framework(self, tuner):
+        """
+        Saves the model weights, optimizer, scheduler and epoch
+        depending on its framework.
+        """
+        if get_framework(tuner.embed_model) == 'keras':
+            tuner.save(
+                filepath=self._get_file_path(tuner),
+            )
+            state = {
+                'epoch': tuner.state.epoch + 1,
+            }
+            with open(
+                os.path.join(self._get_file_path(tuner), 'saved_state.pkl'), 'wb'
+            ) as f:
+                pickle.dump(state, f)
+        elif get_framework(tuner.embed_model) == 'torch':
+            if tuner._scheduler and hasattr(tuner._scheduler, 'state_dict'):
+                state = {
+                    'epoch': tuner.state.epoch + 1,
+                    'state_dict': tuner.embed_model.state_dict(),
+                    'optimizer': tuner._optimizer.state_dict(),
+                    'scheduler': tuner._scheduler.state_dict(),
+                }
+            else:
+                state = {
+                    'epoch': tuner.state.epoch + 1,
+                    'state_dict': tuner.embed_model.state_dict(),
+                    'optimizer': tuner._optimizer.state_dict(),
+                    'scheduler': None,
+                }
```

Comment:
I would refactor this to create a dict with epoch, state dict and optimizer first, and then add the scheduler to it only if it exists - will be clearer and less code